mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -334,33 +363,84 @@ pub(crate) async fn block_header_utxo_loop<T: UtxoCommonOps>(
                     "Error {error:?} on retrieving latest {ticker} headers from rpc after {FETCH_BLOCK_HEADERS_ATTEMPTS} attempts"
                 );
                 // Todo: remove this electrum server and use another in this case since the headers from this server can't be retrieved
-                sync_status_loop_handle.notify_on_permanent_error(error.to_string());
+                sync_status_loop_handle.notify_on_permanent_error(error);
                 break;
             },
         };
 
-        // Validate retrieved block headers
-        if let Some(params) = &coin.as_ref().conf.block_headers_verification_params {
-            if let Err(e) = validate_headers(ticker, from_block_height, block_headers, storage, params).await {
-                error!("Error {e:?} on validating the latest headers for {ticker}!");
-                // Todo: remove this electrum server and use another in this case since the headers from this server are invalid
-                sync_status_loop_handle.notify_on_permanent_error(e.to_string());
-                break;
-            }
+        // Validate retrieved block headers.
+        if let Err(err) = validate_headers(ticker, from_block_height, block_headers, storage, &spv_conf).await {
+            error!("Error {err:?} on validating the latest headers for {ticker}!");
+            // Todo: remove this electrum server and use another in this case since the headers from this server are invalid
+            sync_status_loop_handle.notify_on_permanent_error(err);
+            break;
         };
 
         let sleep = args.success_sleep;
         ok_or_continue_after_sleep!(storage.add_block_headers_to_storage(block_registry).await, sleep);
     }
 }
 
+async fn save_starting_block_header_to_storage(
```

Comment:
Please consider the following changes
```diff
Subject: [PATCH] update rust-lightning to v0.0.113 wip
---
Index: mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs b/mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs
--- a/mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs	(revision f957ddc96ba9a6d038c0dc3713233e31eaa7bffa)
+++ b/mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs	(date 1674755001935)
@@ -1,4 +1,4 @@
-use crate::utxo::rpc_clients::{ElectrumClient, UtxoRpcClientEnum, UtxoRpcError};
+use crate::utxo::rpc_clients::{ElectrumClient, UtxoRpcClientEnum};
 use crate::utxo::utxo_block_header_storage::BlockHeaderStorage;
 use crate::utxo::utxo_builder::{UtxoCoinBuildError, UtxoCoinBuilder, UtxoCoinBuilderCommonOps,
                                 UtxoFieldsWithGlobalHDBuilder, UtxoFieldsWithHardwareWalletBuilder,
@@ -268,19 +268,12 @@
         let storage = client.block_headers_storage();
         let from_block_height = match storage.get_last_block_height().await {
             Ok(Some(height)) => height,
-            Ok(None) => match save_starting_block_header_to_storage(client, ticker, storage, &spv_conf).await {
-                Ok(height) => height,
-                Err(err) => {
-                    if let BlockHeaderStorageError::BadBlockHeader { .. } = &err {
-                        error!("{err:?}");
-                        sync_status_loop_handle.notify_on_permanent_error(err);
-                        continue;
-                    };
-
-                    error!("{err:?}");
-                    sync_status_loop_handle.notify_on_temp_error(err);
-                    continue;
-                },
+            Ok(None) => {
+                if let Err(e) = validate_and_store_starting_header(client, ticker, storage, &spv_conf).await {
+                    sync_status_loop_handle.notify_on_permanent_error(e);
+                    break;
+                }
+                spv_conf.starting_block_header.height
             },
             Err(err) => {
                 error!("Error {err:?} on getting the height of the last stored {ticker} header in DB!",);
@@ -381,42 +374,52 @@
     }
 }
 
-async fn save_starting_block_header_to_storage(
+#[derive(Display)]
+enum StartingHeaderValidationError {
+    #[display(fmt = "Can't decode/deserialize from storage for {} - reason: {}", coin, reason)]
+    DecodeErr {
+        coin: String,
+        reason: String,
+    },
+    RpcError(String),
+    StorageError(String),
+    #[display(fmt = "Error validating starting header for {} - reason: {}", coin, reason)]
+    ValidationError {
+        coin: String,
+        reason: String,
+    },
+}
+
+async fn validate_and_store_starting_header(
     client: &ElectrumClient,
     ticker: &str,
     storage: &BlockHeaderStorage,
     spv_conf: &SPVConf,
-) -> Result<u64, BlockHeaderStorageError> {
+) -> Result<(), StartingHeaderValidationError> {
     let height = spv_conf.starting_block_header.height;
-
-    match client
+    let header_bytes = client
         .blockchain_block_header(height)
-        .map_to_mm_fut(UtxoRpcError::from)
         .compat()
         .await
-    {
-        Ok(header_bytes) => {
-            let mut reader = Reader::new(&header_bytes);
-            let header = reader.read().map_err(|err| BlockHeaderStorageError::DecodeError {
-                coin: ticker.to_string(),
-                reason: err.to_string(),
-            })?;
+        .map_err(|e| StartingHeaderValidationError::RpcError(e.to_string()))?;
+
+    let mut reader = Reader::new_with_coin_variant(&header_bytes, ticker.into());
+    let header = reader.read().map_err(|err| StartingHeaderValidationError::DecodeErr {
+        coin: ticker.to_string(),
+        reason: err.to_string(),
+    })?;
 
-            spv_conf
-                .validate_rpc_starting_block_header(height, &header)
-                .map_err(|err| BlockHeaderStorageError::BadBlockHeader {
-                    coin: ticker.to_string(),
-                    reason: err.to_string(),
-                })?;
+    spv_conf.validate_rpc_starting_header(height, &header).map_err(|err| {
+        StartingHeaderValidationError::ValidationError {
+            coin: ticker.to_string(),
+            reason: err.to_string(),
+        }
+    })?;
 
-            storage
-                .add_block_headers_to_storage(HashMap::from([(height, header)]))
-                .await?;
-
-            Ok(height)
-        },
-        Err(err) => Err(BlockHeaderStorageError::JsonRpcError(err.to_string())),
-    }
+    storage
+        .add_block_headers_to_storage(HashMap::from([(height, header)]))
+        .await
+        .map_err(|e| StartingHeaderValidationError::StorageError(e.to_string()))
 }
 
 async fn remove_excessive_headers_from_storage(
Index: mm2src/mm2_bitcoin/spv_validation/src/storage.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/mm2_bitcoin/spv_validation/src/storage.rs b/mm2src/mm2_bitcoin/spv_validation/src/storage.rs
--- a/mm2src/mm2_bitcoin/spv_validation/src/storage.rs	(revision f957ddc96ba9a6d038c0dc3713233e31eaa7bffa)
+++ b/mm2src/mm2_bitcoin/spv_validation/src/storage.rs	(date 1674754010647)
@@ -37,18 +37,12 @@
         coin: String,
         reason: String,
     },
-    #[display(fmt = "Bad Block Header  - coin: {} - reason: {}", coin, reason)]
-    BadBlockHeader {
-        coin: String,
-        reason: String,
-    },
     #[display(fmt = "Can't decode/deserialize from storage for {} - reason: {}", coin, reason)]
     DecodeError {
         coin: String,
         reason: String,
     },
     Internal(String),
-    JsonRpcError(String),
 }
 
 #[async_trait]
Index: mm2src/mm2_bitcoin/spv_validation/src/conf.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/mm2_bitcoin/spv_validation/src/conf.rs b/mm2src/mm2_bitcoin/spv_validation/src/conf.rs
--- a/mm2src/mm2_bitcoin/spv_validation/src/conf.rs	(revision f957ddc96ba9a6d038c0dc3713233e31eaa7bffa)
+++ b/mm2src/mm2_bitcoin/spv_validation/src/conf.rs	(date 1674754482100)
@@ -53,7 +53,7 @@
 /// Validate that `max_stored_headers_value` is always greater than `retarget interval`.
 fn validate_btc_max_stored_headers_value(max_stored_block_headers: u64) -> Result<(), SPVError> {
     if RETARGETING_INTERVAL > max_stored_block_headers as u32 {
-        return Err(SPVError::SPVBlockHeaderError(format!(
+        return Err(SPVError::InitialValidationError(format!(
             "max_stored_block_headers {max_stored_block_headers} must be greater than retargeting interval {RETARGETING_INTERVAL}",
         )));
     }
@@ -116,7 +116,7 @@
     }
 
     /// Validate Starting block header from `RPC` against [`SPVConf::SPVBlockHeader`]
-    pub fn validate_rpc_starting_block_header(&self, height: u64, rpc_header: &BlockHeader) -> Result<(), SPVError> {
+    pub fn validate_rpc_starting_header(&self, height: u64, rpc_header: &BlockHeader) -> Result<(), SPVError> {
         let rpc_header = SPVBlockHeader::from_block_header_and_height(rpc_header, height);
         let spv_header = &self.starting_block_header;
 
@@ -125,21 +125,21 @@
         let rpc_header_bits: u32 = rpc_header.bits.into();
         let spv_header_bits: u32 = spv_header.bits.clone().into();
         if rpc_header_bits != spv_header_bits {
-            return Err(SPVError::SPVBlockHeaderError(format!(
-                "Block header bits not acceptable - expected: {spv_header_bits} - found: {rpc_header_bits}"
+            return Err(SPVError::InitialValidationError(format!(
+                "Starting block header bits not acceptable - expected: {spv_header_bits} - found: {rpc_header_bits}"
             )));
         };
 
         if rpc_header.hash != spv_header.hash {
-            return Err(SPVError::SPVBlockHeaderError(format!(
-                "Block header hash not acceptable - expected: {} - found: {}",
+            return Err(SPVError::InitialValidationError(format!(
+                "Starting block header hash not acceptable - expected: {} - found: {}",
                 spv_header.hash, rpc_header.hash,
             )));
         };
 
         if rpc_header.time != spv_header.time {
-            return Err(SPVError::SPVBlockHeaderError(format!(
-                "Block header time not acceptable - expected: {} - found: {}",
+            return Err(SPVError::InitialValidationError(format!(
+                "Starting block header time not acceptable - expected: {} - found: {}",
                 spv_header.time, rpc_header.time,
             )));
         };
Index: mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs b/mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs
--- a/mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs	(revision f957ddc96ba9a6d038c0dc3713233e31eaa7bffa)
+++ b/mm2src/mm2_bitcoin/spv_validation/src/helpers_validation.rs	(date 1674755002220)
@@ -10,8 +10,8 @@
 
 #[derive(Clone, Debug, Display, Eq, PartialEq)]
 pub enum SPVError {
-    #[display(fmt = "Starting block header error: {_0}")]
-    SPVBlockHeaderError(String),
+    #[display(fmt = "Error validating initial spv parameters: {_0}")]
+    InitialValidationError(String),
     #[display(fmt = "Overran a checked read on a slice")]
     ReadOverrun,
     #[display(fmt = "Attempted to parse a CompactInt without enough bytes")]
@@ -54,13 +54,13 @@
     Timeout,
     #[display(fmt = "Block headers storage error: {}", _0)]
     HeaderStorageError(BlockHeaderStorageError),
-    #[display(fmt = "Internal error: {}", _0)]
-    Internal(String),
     #[display(
-        fmt = "Wrong retarget block header height in config for: {coin} expected block header height : 
+        fmt = "Wrong retarget block header height in config for: {coin} expected block header height :
         {expected_height}."
     )]
     WrongRetargetHeight { coin: String, expected_height: u64 },
+    #[display(fmt = "Internal error: {}", _0)]
+    Internal(String),
 }
 
 impl From<RawHeaderError> for SPVError {
```
- You should always use `Reader::new_with_coin_variant` when deserializing headers.
- `JsonRpcError` shouldn't be a variant inside `BlockHeadersStorageError`.
- `SPVBlockHeaderError` is not a good error variant name since `SPVBlockHeader` is  also used for other block headers not only the starting one.
- `save_starting_block_header_to_storage` is not indicative of what the function does, it's renamed to `validate_and_store_starting_header`. Also this function shouldn't return a value as implied by it's name.
- Permanent error should be returned when there is a problem in validating or storing the starting error, since the only temporary error that can happen is related to rpc and it's good to also return this error in the activation process if the electrum server/s is not responding.

Reply:
ok noted thank you!